File: 303.py
class NumArray:

    def __init__(self, nums: List[int]):
        temSum = 0
        self.sums = []
        for i in nums:
            temSum += i
            self.sums.append(temSum)

    def sumRange(self, i: int, j: int) -> int:
        if i == 0:
            return self.sums[j]
        return self.sums[j] - self.sums[i-1]

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)

File: 303.js
/*
 * @lc app=leetcode.cn id=303 lang=javascript
 *
 * [303] 区域和检索 - 数组不可变
 */

// @lc code=start
/**
 * @param {number[]} nums
 */
var NumArray = function (nums) {
  this.pre_sum = [0];
  nums.forEach(val=>{
    this.pre_sum.push(val + this.pre_sum[this.pre_sum.length-1])
  })
};

/**
 * @param {number} i
 * @param {number} j
 * @return {number}
 */
NumArray.prototype.sumRange = function (i, j) {
  return this.pre_sum[j+1] - this.pre_sum[i];
};

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(i,j)
 */
// @lc code=end
new NumArray([-2, 0, 3, -5, 2, -1]);
